File: mage/src/main/java/mil/nga/giat/mage/coordinate/CoordinateFormatter.java
package mil.nga.giat.mage.coordinate;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.android.gms.maps.model.LatLng;

import java.text.DecimalFormat;

import mil.nga.giat.mage.R;
import mil.nga.mgrs.MGRS;

public class CoordinateFormatter {
    private CoordinateSystem coordinateSystem;
    private DecimalFormat latLngFormat = new DecimalFormat("###.00000");

    public CoordinateFormatter(Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        int preferenceValue = preferences.getInt(context.getString(R.string.coordinateSystemViewKey), context.getResources().getInteger(R.integer.coordinateSystemViewDefaultValue));
        coordinateSystem = CoordinateSystem.get(preferenceValue);
    }

    public CoordinateSystem getCoordinateSystem() {
        return coordinateSystem;
    }

    public String format(LatLng latLng) {
        switch(coordinateSystem) {
            case MGRS:
                return getMGRS(latLng);
            default:
                return getWGS84(latLng);
        }
    }

    private String getWGS84(LatLng latLng) {
        return latLngFormat.format(latLng.latitude) + ", " + latLngFormat.format(latLng.longitude);
    }

    private String getMGRS(LatLng latLng) {
        MGRS mgrs = MGRS.from(new mil.nga.mgrs.wgs84.LatLng(latLng.latitude, latLng.longitude));
        return mgrs.format(5);
    }
}

File: mage/src/main/java/mil/nga/giat/mage/coordinate/CoordinateFormatter.kt
package mil.nga.giat.mage.coordinate

import android.content.Context
import androidx.preference.PreferenceManager
import com.google.android.gms.maps.model.LatLng
import mil.nga.gars.GARS
import mil.nga.giat.mage.R
import mil.nga.mgrs.MGRS
import java.text.DecimalFormat

class CoordinateFormatter(context: Context) {
   private val coordinateSystem: CoordinateSystem
   private val latLngFormat = DecimalFormat("###.00000")

   fun format(latLng: LatLng): String {
      return when (coordinateSystem) {
         CoordinateSystem.WGS84 -> {
            latLngFormat.format(latLng.latitude) + ", " + latLngFormat.format(latLng.longitude)
         }
         CoordinateSystem.MGRS -> {
            val mgrs = MGRS.from(mil.nga.grid.features.Point.point(latLng.longitude, latLng.latitude))
            mgrs.coordinate()
         }
         CoordinateSystem.DMS -> {
            DMS.from(latLng).format()
         }
         CoordinateSystem.GARS -> {
            val gars = GARS.from(mil.nga.grid.features.Point.point(latLng.longitude, latLng.latitude))
            gars.coordinate()
         }
      }
   }

   init {
      val value = PreferenceManager.getDefaultSharedPreferences(context).getInt(
         context.getString(R.string.coordinateSystemViewKey),
         context.resources.getInteger(R.integer.coordinateSystemViewDefaultValue)
      )

      coordinateSystem = CoordinateSystem.fromPreference(value)
   }
}